## Epigenetic heterogeneity and transcriptional drivers of triple-negative breast cancer
This repository contains downstream analysis code from our study "Epigenetic heterogeneity and transcriptional drivers of triple-negative breast cancer"

## Abstract
Triple-negative breast cancer (TNBC) is a heterogeneous disease with limited treatment options. To characterize TNBC heterogeneity, we defined transcriptional, epigenetic, and metabolic subtypes, and subtype-driving super-enhancers and transcription factors by combining functional and molecular profiling with computational analyses. Single cell RNA sequencing revealed relative homogeneity of the major transcriptional subtypes (luminal, basal, and mesenchymal) within samples. We found that mesenchymal TNBCs are more similar to mesenchymal neuroblastoma and rhabdoid tumors than to other TNBC subtypes, and that the PRRX1 transcription factor is a key driver of these tumors. PRRX1 is sufficient for inducing mesenchymal features in basal but not in luminal TNBC cells via reprogramming super-enhancer landscapes but it is not required for mesenchymal state maintenance or for cellular viability. Our comprehensive, large scale, multi-platform, multi-omics study of both experimental and clinical TNBC is an important resource for the scientific and clinical research communities and opens new venues for future investigation. 

## Content
* `/scripts/`: Downstream analysis code used to create the figures and run the analyses presented in the paper
* `/utilities/`: Utilities scripts used by the analysis code
* `/analysis/`: Output directory
* `/R_Data/`: Directory for preprocessed .RData files

## Data
Raw and processed NGS and methylation array data are available from GEO (accession: GSE202776). Additional metadata and processed data files required to run the code are available to download from Figshare.

## Contact
For any questions, please feel free to reach out to me at dtemko@jimmy.harvard.edu.
